src/Middleware/TaskLockBagMiddleware.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SchedulerBundle\Middleware;
+
+use DateTimeImmutable;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use SchedulerBundle\SchedulerInterface;
+use SchedulerBundle\Task\TaskInterface;
+use SchedulerBundle\TaskBag\LockTaskBag;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\LockFactory;
+use Throwable;
+
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+final class TaskLockBagMiddleware implements PostSchedulingMiddlewareInterface
+{
+    public const TASK_LOCK_MASK = '_symfony_scheduler_';
+
+    private LockFactory $lockFactory;
+    private ?LoggerInterface $logger;
+
+    public function __construct(
+        LockFactory $lockFactory,
+        ?LoggerInterface $logger = null
+    ) {
+        $this->lockFactory = $lockFactory;
+        $this->logger = $logger ?? new NullLogger();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function postScheduling(TaskInterface $task, SchedulerInterface $scheduler): void
+    {
+        if (null !== $task->getExecutionLockBag()) {
+            return;
+        }
+
+        $key = new Key(sprintf('%s_%s_%s', self::TASK_LOCK_MASK, $task->getName(), (new DateTimeImmutable())->format($task->isSingleRun() ? 'Y_m_d_h' : 'Y_m_d_h_i')));
+
+        $lock = $this->lockFactory->createLockFromKey($key, null, false);
+        if (!$lock->acquire(true)) {
+            return;
+        }
+
+        try {
+            $scheduler->update($task->getName(), $task->setExecutionLockBag(new LockTaskBag($key)));
+        } catch (Throwable $throwable) {
+            $this->logger->critical(sprintf('The lock for the task "%s" cannot be serialized / stored, consider using a supporting store', $task->getName()));
+
+            $lock->release();
+        } finally {
+            $scheduler->update($task->getName(), $task->setExecutionLockBag());
```

Comment:
what does this finally block correspond to?

should it not be called only when the try update failed? because there, if I understand correctly it overwrites the update of the try block no?

Reply:
You're right, the `finally` block override the default one, fixed 🙂